src/main/java/com/direwolf20/buildinggadgets/common/schema/template/Template.java

```diff
@@ -0,0 +1,241 @@
+    private static final String KEY_TILE_DATA = "tile_data";
+    private static final String KEY_AUTHOR = "author";
+    private static final int BYTE_MASK_12BIT = 0xF_FF;
+    private static final int BYTE_MASK_14BIT = 0x3F_FF;
+    private static final int BYTE_MASK_24BIT = 0xFF_FF_FF;
+    private static final long BYTE_MASK_40BIT = ((long) 0xFF_FF_FF_FF) << 8 | 0xFF;
+    private static final int MAX_NUMBER_OF_STATES = (1 << 24) - 1;
+    private final Long2ObjectMap<BlockState> states;
+    private final BoundingBox bounds;
+    private final String author;
+
+    public static Optional<Template> deserializeNBT(CompoundNBT nbt) {
+        if (! nbt.contains(KEY_POS, NBT.TAG_LONG_ARRAY) || ! nbt.contains(KEY_DATA, NBT.TAG_LIST))
+            return Optional.empty();
+
+        Long2ObjectMap<BlockState> states = new Long2ObjectLinkedOpenHashMap<>();
+        BoundingBox bounds = null;
+        String author = nbt.getString(KEY_AUTHOR);
+
+        //read the states
+        ListNBT nbtStates = (ListNBT) nbt.get(KEY_DATA);
+        assert nbtStates != null;
+        List<BlockState> uniqueStates = new ArrayList<>(nbtStates.size());
+        for (INBT stateNBT : nbtStates) {
+            //read blockstate will replace everything unknown with the default value for Blocks: AIR
+            BlockState state = NBTUtil.readBlockState((CompoundNBT) stateNBT);
+            if (state == Blocks.AIR.getDefaultState())
+                //a trace log, as this may be normal... In the future we might want to replace this with some placeholder
+                BuildingGadgets.LOGGER.trace("Found unknown state with nbt {}. Will be ignored!", stateNBT);
+            uniqueStates.add(state);
+        }
+        //process the state map
+        int minX = 0, minY = 0, minZ = 0, maxX = 0, maxY = 0, maxZ = 0;
+        Mutable thePos = new Mutable();
+        for (long packed : nbt.getLongArray(KEY_POS)) {
+            int id = getId(packed);
+            if (id >= uniqueStates.size() || id < 0) {
+                BuildingGadgets.LOGGER.error("Found illegal state-id {} in Template at pos {}(={}), when {} states " +
+                                "are known. This indicates a greater Problem and the Template is therefore deemed broken.",
+                        id, getPos(packed), posFromLong(new Mutable(), getPos(packed)), uniqueStates.size());
+                return Optional.empty();
+            }
+            BlockState state = uniqueStates.get(id);
+            if (state == Blocks.AIR.getDefaultState())
+                //Skip unknown Blocks - this has no log, as it is already logged before and even might happen more
+                //than once. So having a log here might result in serious log spam
+                continue;
+            long posLong = getPos(packed);
+            posFromLong(thePos, posLong);
+            //I'd actually like to outsource this to a class, but Mike told me not to...
+            minX = Math.min(thePos.getX(), minX);
```

Comment:
Is this not just this: https://s.mikey.pro/i/84ea19d616.png

Reply:
No, because I'm updating the positions. Normally I would have done this with the Builder, but as I'm not supposed to write one I need to update the variables.